scheduler/reconcile_util.go

```diff
@@ -199,33 +199,46 @@ func (a allocSet) filterByTainted(nodes map[string]*structs.Node) (untainted, mi
 	migrate = make(map[string]*structs.Allocation)
 	lost = make(map[string]*structs.Allocation)
 	for _, alloc := range a {
-		n, ok := nodes[alloc.NodeID]
-		if !ok {
+		// Terminal allocs are always untainted as they should never be migrated
+		if alloc.TerminalStatus() {
 			untainted[alloc.ID] = alloc
 			continue
 		}
 
+		// Non-terminal allocs that should migrate should always migrate
+		if alloc.DesiredTransition.ShouldMigrate() {
+			migrate[alloc.ID] = alloc
+			continue
+		}
+
+		//FIXME is this neccessary?! Surely RanSuccessfully() is only
+		//      true if the alloc is Terminal?
+
 		// If the job is batch and finished successfully, the fact that the
 		// node is tainted does not mean it should be migrated or marked as
 		// lost as the work was already successfully finished. However for
 		// service/system jobs, tasks should never complete. The check of
 		// batch type, defends against client bugs.
-		if alloc.Job.Type == structs.JobTypeBatch && alloc.RanSuccessfully() {
+		if alloc.Job != nil && alloc.Job.Type == structs.JobTypeBatch && alloc.RanSuccessfully() {
 			untainted[alloc.ID] = alloc
 			continue
 		}
 
-		if !alloc.TerminalStatus() {
-			if n == nil || n.TerminalStatus() {
-				lost[alloc.ID] = alloc
-			} else if alloc.DesiredTransistion.ShouldMigrate() {
-				migrate[alloc.ID] = alloc
-			} else {
-				untainted[alloc.ID] = alloc
-			}
-		} else {
+		n, ok := nodes[alloc.NodeID]
+		if !ok {
+			//FIXME why are unknown nodes untainted?! shouldn't this be "lost"?!
```

Comment:
The passed in map is only tainted nodes so if it isn't in there it is untainted.